devito/operator.py

```diff
@@ -96,6 +96,17 @@ def expr_cse(expr):
                     s_dict[arg] = Indexed(to_revert[value.base.label], *value.indices)
         to_revert[temp] = value.xreplace(s_dict)
 
+    # Make sure that no temporary that has to be reverted depends
+    # on another temporary of the list
+    changing = True
+    while changing:
+        changing = False
+        for temp, value in to_revert.items():
+            n_value = value.xreplace(to_revert)
```

Comment:
Ok I have a couple ideas. 

Also this needs to be fixed since lots of things that you get in `to_revert` shouldn't be there. I need a better way of reverting only stuff that is used in indexes